### PR TITLE
Propose addition of PSL catalog file

### PR DESCRIPTION
**Pull request by hdoupe**
This PR adds a configuration file for the [PSL-Core catalog]. This file is used by the Catalog-Builder tool to locate data that is necessary to build the catalog. Documentation describing how the builder works can be found in the [PSL-Core Catalog-Builder REAMDE.md].

I've collected most of the required attributes, but the maintainers of Tax-Calculator may want to use different information. Attributes that I was unable to find or fill in are "key features", "user changelog", "user case studies", "governance overview", "public funding", and "project roadmap". I'm happy to help curate this information if it does not already exist in Tax-Calculator.

@MattHJensen @martinholmer would you mind reviewing this PR and helping to fill in the remaining attributes?

Also, if there are objections to adding this file to the Tax-Calculator repository, then I'm happy to discuss them. The primary rationale for putting these files in the upstream project repository was that it would make it easier for the upstream project maintainers to update their project's cataloged information.

[EDIT] Here's a [link to the Tax-Calculator page](https://open-source-economics.github.io/PSL-Core/Core/Tools/Web/pages/projects/Tax-Calculator.html) on the PSL-Core website.

[PSL-Core catalog]: https://github.com/open-source-economics/PSL-Core/tree/master/Catalog
[PSL-Core Catalog-Builder REAMDE.md]: https://github.com/open-source-economics/PSL-Core/tree/master/Core/Tools/Catalog-Builder

### CODE REVIEW

**Comment by martinholmer** [comment]
@hdoupe said in PR #2082:
> This PR adds a configuration file for the [PSL-Core catalog](https://github.com/open-source-economics/PSL-Core/tree/master/Catalog). This file is used by the Catalog-Builder tool to locate data that is necessary to build the catalog. Documentation describing how the builder works can be found in the [PSL-Core Catalog-Builder REAMDE.md](https://github.com/open-source-economics/PSL-Core/blob/master/Core/Tools/Catalog-Builder/README.md).

Hank,  I've got a couple questions about this.

1.  Why does the PR title and text (see above) refer to the new file as a "configuration" file?  Doesn't seem like there is any configuration going on here.  And when I read the `README.md` file you link to, I don't see any mention of configuration files.  Maybe you need a better name for these kinds of files.  Why not just call them catalog files?

2. Why is the file named `psl_catalog.json` and not `PSL_catalog.json`?  I thought that PSL was an acronym for Policy Simulation Library.  Aren't acronyms capitalized?

Also, after glancing over the linked `README.md` file and reading the rest of your PR comment, I'm confused about the status of the keys in the catalog file.  The `README.md` file describes the catalog file keys as "Allowed attributes," which strongly suggests that they are all optional.  So, why do you say latter in your PR comment:
> Attributes that I was unable to find or fill in are "key features", "user changelog", "user case studies", "governance overview", "public funding", and "project roadmap". I'm happy to help curate this information if it does not already exist in Tax-Calculator.

That statement makes it sound like all the information is **manditory**.  But if that is the case, I don't remember having a conversation about which PSL catalog keys are manditory and which catalog keys are optional.

**Comment by martinholmer** [comment]
@hdoupe, This is a minor problem in PR #2082: the proposed JSON file is missing a newline at the end:
```
Tax-Calculator$ pycodestyle --ignore=E501,E121 *.json
psl_catalog.json:156:2: W292 no newline at end of file
```

**Comment by martinholmer** [comment]
@hdoupe proposed a new PSL catalog file that contain this entry:
```
    "contributor_guide": {
        "start_header": "Tax-Calculator Coding Style",
        "end_header": "pycodestype (the program formerly known as pep8)",
        "source": "CODING.md",
        "type": "github_file",
        "data": null
    },
```
First, there is no `CODING.md` file in the Tax-Calculator repository, but there is a contributor guide.
Why not extract information from that?  Anyway, "coding style" is way too narrow an interpretation of what a contributor guide would be.

**Comment by martinholmer** [comment]
@hdoupe,  It would help people assess your proposal in PR #2082, if you would show them the PSL Catalog entry for Tax-Calculator that the proposed JSON file would generate.  It will be much more readable that the raw JSON.

**Comment by hdoupe** [comment]
@martinholmer Thank you for your extensive feedback on #2082. I will do my best to answer your questions and clear things up in the comments below.

**Comment by hdoupe** [comment]
@martinholmer  asked:
> Why does the PR title and text (see above) refer to the new file as a "configuration" file? Doesn't seem like there is any configuration going on here. And when I read the README.md file you link to, I don't see any mention of configuration files. Maybe you need a better name for these kinds of files. Why not just call them catalog files?

These are all good points. I was struggling to come up with a naming scheme. I think calling it a "catalog file" is fine.

> Why is the file named psl_catalog.json and not PSL_catalog.json? I thought that PSL was an acronym for Policy Simulation Library. Aren't acronyms capitalized?

Yes, but I used lowercase because I tend to do all lowercase letters in file names. I don't think there's a real reason for this though. I'm fine with changing this if it will make the purpose of the file clearer.

**Comment by hdoupe** [comment]
@martinholmer  stated:
> That statement makes it sound like all the information is manditory. But if that is the case, I don't remember having a conversation about which PSL catalog keys are manditory and which catalog keys are optional.

@MattHJensen did we decide that all of the catalog keys in the `psl_catalog.json` file are mandatory?

**Comment by hdoupe** [comment]
@martinholmer I updated the first comment to have a link to the PSL-Core home page. Here it is again for reference: https://open-source-economics.github.io/PSL-Core/Core/Tools/Web/pages/projects/Tax-Calculator.html

**Comment by martinholmer** [comment]
@hdoupe said in PR #2082:
>  I think calling it a "catalog file" is fine.
>
>>Why is the file named psl_catalog.json and not PSL_catalog.json? I thought that PSL was an acronym for Policy Simulation Library. Aren't acronyms capitalized?
>
> Yes, but I used lowercase because I tend to do all lowercase letters in file names. I don't think there's a real reason for this though. I'm fine with changing this if it will make the purpose of the file clearer.

Well, yes, I do thing `PSL_catalog.json` is clearer because PSL is more likely than `psl` to be interpreted as the acronym of the Policy Simulation Library.

**Comment by martinholmer** [comment]
@hdoupe said in PR #2082:
> I updated the first comment to have a link to the PSL-Core home page. Here it is again for reference: https://open-source-economics.github.io/PSL-Core/Core/Tools/Web/pages/projects/Tax-Calculator.html

Thanks for the link to the generated PSL Catalog "card" for Tax-Calculator.  That helps a lot in understanding what the raw JSON does and how the `PSL_catalog.json` file can be improved.  I'm happy to start using it to suggest changes, except there is one problem: it seems to be out-of-date.  (I'm saying that because it contains text from a file (`CODING.md`) that is no longer in the Tax-Calculator repository.)   Can you update the catalog "card" using the current version of the Tax-Calculator repo?  Thanks.

**Comment by hdoupe** [comment]
Sure, I'm glad it helped. I'm working to make the updates that you suggested, including the JSON code style fix.

**Comment by hdoupe** [comment]
@Peter-Metz is going to work on a couple of the changes that @martinholmer suggested. I gave him write access to my fork. So his commits should just pop up on this PR.

**Comment by hdoupe** [comment]
Thanks for pushing the update @Peter-Metz. 

@martinholmer are you OK with displaying a link to the contributor guide page in place of the data from what used to be the `CODING.md` file?

File: PSL_catalog.json [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        "type": "html"
+    },
+    "integration_test": {
+        "start_header": null,
+        "end_header": null,
+        "data": "<a href=\"https://github.com/open-source-economics/Tax-Calculator/tree/master/taxcalc/tests\">https://github.com/open-source-economics/Tax-Calculator/tree/master/taxcalc/tests</a>",
+        "source": "https://github.com/open-source-economics/Tax-Calculator/tree/master/taxcalc/tests",
+        "type": "html"
+    }
+}
```
Comment (hdoupe) [inline PSL_catalog.json]: @Peter-Metz can you add a new line at the end of the `psl-catalog.json` file?

Comment (Peter-Metz) [reply PSL_catalog.json]: @hdoupe Sure thing - see updated

Comment (hdoupe) [reply PSL_catalog.json]: Thanks, this looks good to me.

**Comment by martinholmer** [comment]
@hdoupe asked in PR #2082:
> are you OK with displaying a link to the contributor guide page in place of the text from what used to be the `CODING.md` file?

Yes, that sounds reasonable.

**Comment by martinholmer** [comment]
In the machine-generated Tax-Calculator catalog card, there is this entry:

screen shot 2018-10-15 at 4 08 47 pm

I have two questions:

1. What is the reason for the [source] link when the text above it **is** the "source"?

2. Why is the title missing and why is the formatting of the text (extracted from the `RELEASES.md` file) incorrect?  Here is what the text looks like on the `RELEASES.md` page (and what I would assume is the intended look for the catalog card):

screen shot 2018-10-15 at 4 13 43 pm

**Comment by hdoupe** [comment]
@martinholmer asked:
> What is the reason for the [source] link when the text above it is the "source"?

This makes it easy for people to follow up on the information that is shown. People may be curious where all of this information is collected. It also makes it easy for PSL-Core developers and library members to check that the information was pulled in and formatted correctly.

> Why is the title missing and why is the formatting of the text (extracted from the RELEASES.md file) incorrect? Here is what the text looks like on the RELEASES.md page (and what I would assume is the intended look for the catalog card):

PSL-Core adopted the approach that it would supply the titles. Perhaps, the upstream package could supply some type of sub-title. In the case that you highlight, the two headers are:
- `2018-09-11 Release 0.21.0 : first release compatible only with Python 3.6`
- `2018-09-06 Release 0.20.3 : LAST RELEASE COMPATIBLE WITH PYTHON 2.7`

All of the text in between those headers was pulled. The headers were excluded. Would it be helpful to have an `"include"` tag that specified whether to leave the first header in or out?

I'm not sure what happened to the formatting. The way the parser works is that it pulls the raw markdown file text, renders it to HTML with the [`markdown`](https://github.com/Python-Markdown/markdown) package, and parses the requested information from that. So there are a couple places where the formatting could have been lost. It does look like the markdown dashes that signal bullet points are still there but weren't rendered to HTML lists.

**Comment by MattHJensen** [comment]
> did we decide that all of the catalog keys in the psl_catalog.json file are mandatory?

Not all of the information should be mandatory. Would it be helpful to add "required", "suggested", and "optional" tags to the `PSL_catalog.json` file itself? We would want these to align with the `MUST`s, `SHOULD`s, and `MAY`s from the [library criteria](https://github.com/open-source-economics/PSL-Core/blob/master/Core/Criteria/library_criteria.md).

**Comment by martinholmer** [comment]
@hdoupe said in PR #2082:
> PSL-Core adopted the approach that it would supply the titles. Perhaps, the upstream package could supply some type of sub-title. In the case that you highlight, the two headers are:
>
>* 2018-09-11 Release 0.21.0 : first release compatible only with Python 3.6
>* 2018-09-06 Release 0.20.3 : LAST RELEASE COMPATIBLE WITH PYTHON 2.7
>
>All of the text in between those headers was pulled. The headers were excluded. 
> Would it be helpful to have an "include" tag that specified whether to leave the first header in or out?

Yes it would be **very helpful** to be able to include the first "header".  The first "header" is often the title of the extracted text, which is often crucial for understanding the extracted text.

**Comment by martinholmer** [comment]
@hdoupe said in PR #2082:
> It does look like the markdown dashes that signal bullet points are still there but weren't rendered to HTML lists [in the machine-generated PSL catalog card].

I hope it is easy to fix this because as it is now the extracted text is close to indecipherable.

**Comment by hdoupe** [comment]
@martinholmer I experimented with the markdown text that you referred to here: https://github.com/open-source-economics/Tax-Calculator/pull/2082#issuecomment-429997766.

It looks like there are some nuances between how GitHub deals with Markdown and how the Python [`markdown`](https://github.com/Python-Markdown/markdown) package deals with markdown. Namely, an extra new line is needed before bullet points are listed out and four spaces are needed instead of the usual 2 spaces to signal a sub-bullet point. I would prefer not to get into the nuances of the "correct" way to write markdown. It seems like there is a lot of flexibility in how you can format a markdown document (e.g. I didn't know you could signal lists with "*"s). This makes it complicated when different markdown parsers use different implementations of the markdown specification.

Here's a script if you want to tinker with different variations on the markdown formatting:

```
import markdown
from bs4 import BeautifulSoup

"""
pip install beautifulsoup4
pip install markdown
"""

olddoc = """
TAX-CALCULATOR RELEASE HISTORY
==============================
Go [here](https://github.com/open-source-economics/Tax-Calculator/pulls?q=is%3Apr+is%3Aclosed)
for a complete commit history.


2018-09-11 Release 0.21.0 : first release compatible only with Python 3.6
-------------------------------------------------------------------------
(last merged pull request is
[#2058](https://github.com/open-source-economics/Tax-Calculator/pull/2058))

**API Changes**
- Require Python 3.6 to run Tax-Calculator source code or conda package
  [[#2058](https://github.com/open-source-economics/Tax-Calculator/pull/2058)
  by Martin Holmer], which requires new `puf.csv` input file (see [taxdata pull request 283](https://github.com/open-source-economics/taxdata/pull/283) for details) with this information:
  * Byte size: 56415698
  * MD5 checksum: 3f1c7c2b16b6394a9148779db992bed1

**New Features**
- None

**Bug Fixes**
- None
"""

html = markdown.markdown(olddoc)
soup = BeautifulSoup(html, "html.parser")

print('<h1>OLD DOC</h1>')
print(soup)

newdoc = """
TAX-CALCULATOR RELEASE HISTORY
==============================
Go [here](https://github.com/open-source-economics/Tax-Calculator/pulls?q=is%3Apr+is%3Aclosed)
for a complete commit history.


2018-09-11 Release 0.21.0 : first release compatible only with Python 3.6
-------------------------------------------------------------------------
(last merged pull request is
[#2058](https://github.com/open-source-economics/Tax-Calculator/pull/2058))

**API Changes**

- Require Python 3.6 to run Tax-Calculator source code or conda package
  [[#2058](https://github.com/open-source-economics/Tax-Calculator/pull/2058)
  by Martin Holmer], which requires new `puf.csv` input file (see [taxdata pull request 283](https://github.com/open-source-economics/taxdata/pull/283) for details) with this information:
  * Byte size: 56415698 *(needs four spaces)*
    * MD5 checksum: 3f1c7c2b16b6394a9148779db992bed1

**New Features**
- None *(extra new line needed)*

**Bug Fixes**

- None
"""

html = markdown.markdown(newdoc)
soup = BeautifulSoup(html, "html.parser")

print('<h1>New DOC</h1>')
print(soup)
```

This can be run like:

`python test.py > test.html`

The `test.html` document can be opened in the browser via `open test.html`.

@martinholmer what do you think would be a good way to resolve this problem?

**Comment by martinholmer** [comment]
@hdoupe,  Do the changes in [PSL PR 22](https://github.com/open-source-economics/PSL-Core/pull/22) resolve [this issue](https://github.com/open-source-economics/Tax-Calculator/pull/2082#issuecomment-429997766)?

**Comment by martinholmer** [comment]
@hdoupe said:
> One way that this could be resolved is to reformat the most recent release section in the way that I described in #2082 (comment). The display of the content would be the same, but the way the raw markdown text is formatted would be slightly different (an extra two spaces for the sub-bullets and an extra new line for the first bullet in each section).
>
> Does this seem reasonable?

No, I don't think this is a reasonable solution.  Why don't you use a tool that is up to the job of translating GitHub-flavored markdown to HTML?  There seem to be plenty of options:

screen shot 2018-10-17 at 3 28 37 pm

**Comment by MattHJensen** [comment]
I agree that it would be costly for all PSL projects to train all of their contributors to use non-GH-flavored-markdown. A different solution, If one exists and is reasonable, seems like it’d be preferable.

**Comment by martinholmer** [comment]
@hdoupe said:
> Neither of them cover this edge case. I encourage you to try them in scripts below.

I don't have the time or interest or expertise to try the scripts.
What exactly is "this edge case"?

**Comment by hdoupe** [comment]
@martinholmer asked:
>  What exactly is "this edge case"?

The edge case is the way that the bullets and sub-bullets are formatted. I was able to find a way to re-format the markdown section so that it would render correctly with the `markdown` package and with github. The required edits are minor and would only need to be applied to one section of markdown.

**Comment by martinholmer** [comment]
@hdoupe said:
> The edge case is the way that the bullets and sub-bullets are formatted. I was able to find a way to re-format the markdown section so that it would render correctly with the markdown package and with github. The required edits are minor and would only need to be applied to one section of markdown.

Can you show us how to write markdown that renders the same as now on GitHub and solves your markdown-to-HTML translation problem.   You never showed use how the other tools did with the translation.  Can you show us those results?

**Comment by hdoupe** [comment]
Just updated the comment with the other Python Markdown parsers, too: https://github.com/open-source-economics/Tax-Calculator/pull/2082#issuecomment-430790102

**Comment by hdoupe** [comment]
I'm going out of town for a week. I've asked @Peter-Metz and @andersonfrailey to finish this up while I'm out of town. I've given both of them write-access to my Tax-Calculator fork. So, they should be able to push directly to this PR by doing: 

```
git remote add hdoupe https://github.com/hdoupe/Tax-Calculator
git fetch hdoupe psl-catalog
git checkout psl-catalog
```

Then you can push the changes by doing: 

```
git push hdoupe psl-catalog
```

**Comment by MattHJensen** [comment]
@hdoupe, are the changes to RELEASES.md the only changes that need to be made in the entire file? They seem pretty minor.

**Comment by hdoupe** [comment]
In https://github.com/open-source-economics/Tax-Calculator/pull/2082#issuecomment-429997766, @martinholmer  asked:

> Why is the title missing and why is the formatting of the text (extracted from the RELEASES.md file) incorrect? Here is what the text looks like on the RELEASES.md page (and what I would assume is the intended look for the catalog card):

Does the most recent commit https://github.com/open-source-economics/Tax-Calculator/pull/2082/commits/d24f94f2c551be0a0f2c55fcf522823bf2a8c0ed help?

screen shot 2018-10-19 at 11 27 29 am

The header size in RELEASES.md is a little big compared to the headers on the PSL project page. The headers in the PSL project page are `<h4>` and the RELEASES.md headers are `<h2>`.

**Comment by MattHJensen** [comment]
I see. I guess the concern for Tax-Calculator developers is that they might push a release without noticing a problem, and then PSL would pull the new catalog file, and then the Tax-Calculator catalog card wouldn't look right for some period of time, and that would reflect poorly on the Tax-Calculator project. 

From my perspective, at least, the problem is not so much that an unanticipated rendering occurs, but that there is not a good way -- so far-- for the Tax-Calculator team to catch the unanticipated rendering before publishing a release (or committing to master, for that matter). 

Do you see a way around this issue? 

One option that comes to mind is for Tax-Calculator devs to run the Catalog-Builder during the release and manually review everything, but that seems pretty burdensome and probably ineffective. 

Another approach might be to assume that the number of cases like this are very small, and put a test in Tax-Calculator (and other projects) that warns the project if their markdown contains something that isn't going to render right. 

Maybe I'm missing some other options.

**Comment by MattHJensen** [comment]
I am highly motivated to get the Tax-Calculator card into the PSL catalog quickly, however, I am inclined to not display RELEASES.md in the PSL catalog until [these issues](https://github.com/open-source-economics/Tax-Calculator/pull/2082#issuecomment-431406079) are resolved. 

I am inclined to follow an approach like this: 

1. Remove RELEASES.md content content from the card in this PR. Merge this PR. 
1. As a stop gap solution, add a test to Tax-Calculator that fails if markdown if the project contains markdown that will cause the catalog to render improperly. Add RELEASES.md content back to the Tax-Calculator card. 
1. Move a discussion of a long term solution to [these issues](https://github.com/open-source-economics/Tax-Calculator/pull/2082#issuecomment-431406079) over to a PSL issue, and get feedback from multiple projects in the library. 
1. Implement long term solution wherever it belongs. 

@Peter-Metz, if others agree with this approach, would you be willing to do (1) and (2). I'd be happy to do (3). 

@martinholmer, does this approach make sense to you? Is there a better alternative?

**Comment by Peter-Metz** [comment]
@MattHJensen Removed RELEASES.md content from the catalog card

**Comment by martinholmer** [comment]
@MattHJensen asked in PR #2082:
> does this approach [dropping the RELEASES.md extraction] make sense to you?

No, it doesn't make sense to me.  It's just kicking the can down the road.

>  Is there a better alternative?

If the available GitHub-markdown-to-HTML translators are not very good (which is what Hank seems to be saying), then relying on **all** PSL developers to write their markdown in a certain way so that the PSL Catalog-Builder will work is a really bad idea.  The problem is that markdown that looks perfectly fine on GitHub will be rendered in an indecipherable way in the PSL Catalog.  Relying on **every** developer to **never** make a mistake (when they can't even see the mistake on GitHub) is just an accident --- actually many accidents --- waiting to happen.

Why not write (as part of the Catalog-Builder) a pre-parser that enforces the markdown format required by the weak translator?  That way the input into the translator will always be "correct" even if PSL model developers forget to write markdown code that is expected by the translator being used in the Catalog-Builder.

@hdoupe @Peter-Metz

**Comment by martinholmer** [comment]
@Peter-Metz said he
> Removed RELEASES.md content from the catalog card

This seems like a very bad idea to me.  See my views in [this comment](https://github.com/open-source-economics/Tax-Calculator/pull/2082#issuecomment-431513162).

**Comment by MattHJensen** [comment]
> Why not write (as part of the Catalog-Builder) a pre-parser that enforces the markdown format required by the weak translator? That way the input into the translator will always be "correct" even if PSL model developers forget to write markdown code that is expected by the translator being used in the Catalog-Builder.

This seems like a very reasonable option. I think Hank deserves to be part of the conversation about how to handle our concerns, though, and he's on vacation for around around 10 days. I really don't want to hold up Tax-Calculator's introduction into PSL for 10 days plus discussion time plus development time, which is why I am suggesting that we "kick the can down the road" for those 10 days. 

Another option would be for @Peter-Metz to take a crack at coding up the pre-parser, and we could use that provisionally, BUT I would only want to do this with the understanding that we would reopen the conversation after Hank is back and can participate.  It would also be helpful if @martinholmer or @andersonfrailey would volunteer ahead of time to review @peter-metz's PR to PSL's Catalog-Builder. 

@martinholmer do you have a preference between the provisional pre-parsing vs. wait and discuss (kick the can) approaches? Are you comfortable w one of these?

**Comment by MattHJensen** [comment]
One other note: If we go the pre-parsing route, even provisionally, I feel pretty strongly that our goal should be narrow: augment the existing weak translator so that we can properly render GitHub flavored markdown. We should be very wary of adding requirements beyond that to the pre parser down the road.

**Comment by martinholmer** [comment]
@MattHJensen said in Tax-Calculator PR #2082:
> Another option would be for @Peter-Metz to take a crack at coding up the pre-parser, and we could use that provisionally, BUT I would only want to do this with the understanding that we would reopen the conversation after Hank is back and can participate. It would also be helpful if @martinholmer or @andersonfrailey would volunteer ahead of time to review @Peter-Metz's PR to PSL's Catalog-Builder.

This seems sensible.  I'm happy to review contributions.

> @martinholmer do you have a preference between the provisional pre-parsing vs. wait and discuss (kick the can) approaches? Are you comfortable w one of these?

Yes, let's try the "provisional pre-parsing" now and see what @hdoupe thinks about it when he returns.

 > One other note: If we go the pre-parsing route, even provisionally, I feel pretty strongly that our goal should be narrow: augment the existing weak translator so that we can properly render GitHub flavored markdown. We should be very wary of adding requirements beyond that to the pre-parser down the road.

I agree completely with this view.

**Comment by Peter-Metz** [comment]
> Another option would be for @Peter-Metz to take a crack at coding up the pre-parser, and we could use that provisionally, BUT I would only want to do this with the understanding that we would reopen the conversation after Hank is back and can participate.

Happy to take a stab at this.

**Comment by andersonfrailey** [comment]
>One other note: If we go the pre-parsing route, even provisionally, I feel pretty strongly that our goal should be narrow: augment the existing weak translator so that we can properly render GitHub flavored markdown. We should be very wary of adding requirements beyond that to the pre parser down the road.

I also agree with this view and will be happy to review the work @Peter-Metz does.

**Comment by hdoupe** [comment]
I'm on board with the provisional pre-parser and setting the narrow goal of properly rendering Github flavored markdown. I'm happy with @Peter-Metz's implementation of the pre-parser in https://github.com/open-source-economics/PSL-Core/pull/27. Is this a satisfactory solution for everyone else, at least for the short-term?

**Comment by martinholmer** [comment]
@hdoupe,  Given all the recent changes in the PSL repository, why is there any need to edit the `RELEASES.md` file in your Tax-Calculator pull request #2082?

**Comment by hdoupe** [comment]
@martinholmer I reverted the changes that I made to `RELEASES.md`. There are still a couple edits on the file because my text editor automatically removes trailing spaces. If you would like, I can reset the file to the master branch.

I like your idea to have an "Earlier Releases:" line. That way the headers in the JSON file will not need to be updated each time there's a Tax-Calculator release.

**Comment by martinholmer** [comment]
@hdoupe said in PR #2082:
> I like your idea to have an "Earlier Releases:" line. That way the headers in the JSON file will not need to be updated each time there's a Tax-Calculator release.

Exactly!

**Comment by martinholmer** [comment]
@hdoupe said in PR #2082:
> I reverted the changes that I made to `RELEASES.md`. There are still a couple edits on the file because my text editor automatically removes trailing spaces. 

That's fine.  Thanks.

**Comment by hdoupe** [comment]
Sure, no problem.

**Comment by martinholmer** [comment]
PR #2082 should not be merged until it is clear how the extracted Tax-Calculator information is going to be used in the PSL repository.  Right now, I don't understand how it is going to be used and I have (as yet) unanswered questions in the PSL repo.

**Pull request merged**